File: solutions/238.Product_of_Array_Except_Self/238.Product_of_Array_Except_Self.py
from ast import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length = len(nums)
        answer, left, right = [0] * length, [0] * length, [0] * length
        left[0] = 1
        right[length - 1] = 1

        for index in range(1, len(nums)):
            left[index] = left[index - 1] * nums[index - 1]

        for i in range(length - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]

        for index, num in enumerate(nums):
            answer[index] = right[index] * left[index]

        return answer

File: solutions/238.Product_of_Array_Except_Self/238.Product_of_Array_Except_Self.ts
function productExceptSelf(nums: number[]): number[] {
  const length = nums.length
  const answer = new Array().fill(length)
  const left = new Array().fill(length)
  const right = new Array().fill(length)
  left[0] = 1
  right[length - 1] = 1

  for(let i = 1; i < length; i++){
      left[i] = left[i - 1] * nums[i - 1]
  }

  for(let i = length - 2; i >= 0; i--){
      right[i] = right[i + 1] * nums[i + 1]
  }

  for(let i = 0; i < length; i++){
      answer[i] = left[i] * right[i] 
  }
  return answer
};